cps/kni.c

```diff
@@ -55,6 +61,239 @@ kni_change_mtu(uint8_t port_id, unsigned new_mtu)
+			kni_name, family, cmd, strerror(errno));
+		return ret;
+	}
+
+	/*
+	 * We specified NLM_F_ACK to get an acknowledgement, so receive the
+	 * ACK and verify that the interface configuration message was valid
+	 * using the default libmnl callback for doing message verification.
+	 */
+
+	ret = mnl_socket_recvfrom(nl, buf, sizeof(buf));
+	if (ret == -1) {
+		RTE_LOG(ERR, GATEKEEPER, "cps: mnl_socket_recvfrom: cannot update %s with new IP address (family %d) (operation %d): %s\n",
+			kni_name, family, cmd, strerror(errno));
+		return ret;
+	}
+
+	ret = mnl_cb_run(buf, ret, seq, portid, NULL, NULL);
+	if (ret == -1) {
+		RTE_LOG(ERR, GATEKEEPER, "cps: mnl_cb_run: cannot update %s with new IP address (family %d) (operation %d): %s\n",
+			kni_name, family, cmd, strerror(errno));
+		return ret;
+	}
+
+	return ret;
+}
+
+static int
+add_ipaddr(struct mnl_socket *nl, int family, void *ipaddr,
+	uint8_t prefixlen, const char *kni_name)
+{
+	return modify_ipaddr(nl, RTM_NEWADDR,
+		NLM_F_CREATE|NLM_F_REQUEST|NLM_F_EXCL, family,
+		ipaddr, prefixlen, kni_name);
+}
+
+static int
+modify_link(struct mnl_socket *nl, struct rte_kni *kni,
+	const char *kni_name, int if_up)
+{
+	char buf[MNL_SOCKET_BUFFER_SIZE];
+	struct nlmsghdr *nlh;
+	struct ifinfomsg *ifm;
+	unsigned int seq, flags = 0;
+	int pid;
+
+	if (if_up)
+		flags |= IFF_UP;
+	else
+		flags &= ~IFF_UP;
+
+	nlh = mnl_nlmsg_put_header(buf);
+	nlh->nlmsg_type	= RTM_NEWLINK;
+	nlh->nlmsg_flags = NLM_F_REQUEST|NLM_F_ACK;
+	nlh->nlmsg_seq = seq = time(NULL);
+
+	ifm = mnl_nlmsg_put_extra_header(nlh, sizeof(*ifm));
+	ifm->ifi_family = AF_UNSPEC;
+	ifm->ifi_change = IFF_UP;
+	ifm->ifi_flags = flags;
+
+	mnl_attr_put_str(nlh, IFLA_IFNAME, kni_name);
+
+	/*
+	 * The KNI library registers callbacks for MTU changes and
+	 * interface up/down events through ethtool. When these
+	 * operations are requested to the kernel through netlink
+	 * messages, they go back to DPDK in userspace when
+	 * rte_kni_handle_request() is invoked. Therefore, if we
+	 * issue this request, we need another process to call
+	 * rte_kni_handle_request() to allow it to proceed.
+	 *
+	 * The DPDK documentation suggests using ifconfig from a
+	 * shell to bring a KNI up after the application starts and
+	 * calls rte_kni_handle_request(). To do so automatically,
+	 * we have to fork a child process to issue the request
+	 * while the parent calls rte_kni_handle_request().
+	 */
+
+	pid = fork();
+	if (pid == -1) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: fork failed, can't modify KNI %s link: %s\n",
+			kni_name, strerror(errno));
+		return -1;
+	} else if (pid == 0) {
+		/*
+		 * Send request to kernel, which will be sent back
+		 * to userspace for the parent to handle.
+		 */
+		int ret = mnl_socket_sendto(nl, nlh, nlh->nlmsg_len);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"cps: mnl_socket_sendto: cannot bring KNI %s %s: %s\n",
+				kni_name, if_up ? "up" : "down",
+				strerror(errno));
+			_exit(EXIT_FAILURE);
```

Comment:
I use `_exit()` to avoid the full cleanup done by `exit()`, which could affect the parent. Googling around, it seems like it's common to do but not clear to me whether it's absolutely necessary in this case. I think the worst it could do it here is cause I/O buffers to be flushed twice. [[1]](http://www.unixguide.net/unix/programming/1.1.3.shtml) [[2]](http://stackoverflow.com/questions/5422831/what-is-the-difference-between-using-exit-exit-in-a-conventional-linux-fo)